src/stan/services/sample/hmc_nuts_diag_e_adapt.hpp

```diff
@@ -101,18 +99,130 @@ int hmc_nuts_diag_e_adapt(
+      model, init, init_inv_metric, random_seed, chain, init_radius, num_warmup,
+      num_samples, num_thin, save_warmup, refresh, stepsize, stepsize_jitter,
+      max_depth, delta, gamma, kappa, t0, init_buffer, term_buffer, window,
+      interrupt, logger, init_writer, sample_writer, diagnostic_writer,
+      dummy_metric_writer);
+}
+
+/**
+ * Runs HMC with NUTS with adaptation using diagonal Euclidean metric,
+ * with identity matrix as initial inv_metric and saves adapted tuning
+ * parameters stepsize and inverse metric.
+ *
+ * @tparam Model Model class
+ * @param[in] model Input model (with data already instantiated)
+ * @param[in] init var context for initialization
+ * @param[in] random_seed random seed for the random number generator
+ * @param[in] chain chain id to advance the pseudo random number generator
+ * @param[in] init_radius radius to initialize
+ * @param[in] num_warmup Number of warmup samples
+ * @param[in] num_samples Number of samples
+ * @param[in] num_thin Number to thin the samples
+ * @param[in] save_warmup Indicates whether to save the warmup iterations
+ * @param[in] refresh Controls the output
+ * @param[in] stepsize initial stepsize for discrete evolution
+ * @param[in] stepsize_jitter uniform random jitter of stepsize
+ * @param[in] max_depth Maximum tree depth
+ * @param[in] delta adaptation target acceptance statistic
+ * @param[in] gamma adaptation regularization scale
+ * @param[in] kappa adaptation relaxation exponent
+ * @param[in] t0 adaptation iteration offset
+ * @param[in] init_buffer width of initial fast adaptation interval
+ * @param[in] term_buffer width of final fast adaptation interval
+ * @param[in] window initial width of slow adaptation interval
+ * @param[in,out] interrupt Callback for interrupts
+ * @param[in,out] logger Logger for messages
+ * @param[in,out] init_writer Writer callback for unconstrained inits
+ * @param[in,out] sample_writer Writer for draws
+ * @param[in,out] diagnostic_writer Writer for diagnostic information
+ * @param[in,out] metric_writer Writer for tuning params
+ * @return error_codes::OK if successful
+ */
+template <class Model>
+int hmc_nuts_diag_e_adapt(
+    Model& model, const stan::io::var_context& init, unsigned int random_seed,
+    unsigned int chain, double init_radius, int num_warmup, int num_samples,
+    int num_thin, bool save_warmup, int refresh, double stepsize,
+    double stepsize_jitter, int max_depth, double delta, double gamma,
+    double kappa, double t0, unsigned int init_buffer, unsigned int term_buffer,
+    unsigned int window, callbacks::interrupt& interrupt,
+    callbacks::logger& logger, callbacks::writer& init_writer,
+    callbacks::writer& sample_writer, callbacks::writer& diagnostic_writer,
+    callbacks::structured_writer& metric_writer) {
+  stan::io::dump dmp
+      = util::create_unit_e_diag_inv_metric(model.num_params_r());
+  stan::io::var_context& unit_e_metric = dmp;
+  return hmc_nuts_diag_e_adapt(
```

Comment:
Question: is it necessary to provide this combination (where you don't initialize the metric but do want it written out in the end)?

Just trying to minimize the boilerplate if possible but if the existing calling structure requires it to be so I understand

Reply:
yes, absolutely we want to do this - the default is to start from the unit metric and then let it evolve during adaptation.  this would be the starting point in trying to figure out what a good metric is.

Reply:
Oh, surely, the end user probably wants this option more than many others. What I meant more so was do we actually need this overload provided for everything to compile on our end (e.g., in cmdstan we never call this, just the one with `num_chains`, right?)

Reply:
perhaps only the unit tests will break?  I can test against CmdStan.